src/hash.rs

```diff
@@ -74,13 +94,38 @@ impl HashThing {
     }
 }
 
+#[cfg_attr(feature = "std", derive(Copy, Clone))]
 pub struct ZopfliHash {
     hash1: HashThing,
     hash2: HashThing,
     pub same: [u16; ZOPFLI_WINDOW_SIZE], /* Amount of repetitions of same byte after this .*/
 }
 
+#[cfg(feature = "std")]
+const EMPTY_ZOPFLI_HASH: Lazy<Box<ZopfliHash>> = Lazy::new(|| unsafe {
+    let layout = Layout::new::<ZopfliHash>();
+    let ptr = alloc(layout) as *mut ZopfliHash;
+    for i in 0..ZOPFLI_WINDOW_SIZE {
+        (*ptr).hash1.prev_and_hashval[i].prev = i as u16;
+        (*ptr).hash2.prev_and_hashval[i].prev = i as u16;
+        (*ptr).hash1.prev_and_hashval[i].hashval = None;
+        (*ptr).hash2.prev_and_hashval[i].hashval = None;
+    }
+    (*ptr).hash1.head = [-1; 65536];
+    (*ptr).hash2.head = [-1; 65536];
+    (*ptr).hash1.val = 0;
+    (*ptr).hash2.val = 0;
+    (*ptr).same = [0; ZOPFLI_WINDOW_SIZE];
+    Box::from_raw(ptr)
+});
```

Comment:
This unsafe code is unsound because it does not handle the case where [`alloc` might return a null pointer to indicate allocation failure](https://doc.rust-lang.org/stable/std/alloc/trait.GlobalAlloc.html#tymethod.alloc). However, this function is expected to be called only once, so avoiding `unsafe` altogether seems like an even better approach. What about just using `Box::new`? Were you [concerned about stack usage](https://github.com/rust-lang/rust/issues/63291)?

Reply:
Which of the workarounds discussed in that issue are actually available on the MSRV of 1.62? I got the impression that none of them were.

Reply:
I propose [using boxed slices constructed from `Vec`s instead](https://github.com/rust-lang/rust/issues/63291#issuecomment-1281342953). Given that these arrays are quite large, I see no meaningful cache-friendliness benefit from avoiding the extra indirection, but some numbers to assess this would be great.

Reply:
I figure the flattening of the struct is, more than anything, what explains the performance increase I cited above. I tried a variant of the trick you cited above involving a Vec<ZopfliHash> of length 1, but that still constructs the element on the stack. However, my latest commit checks the allocated pointer for null and converts it to a Box<ZopfliHash> before initializing it, so it should now be sound.

Reply:
That's better!

I'd like to think about how to not use `unsafe` here a bit more though, as relying on it for these cases tends to be a bit of a code smell, indicating that something can be implemented in more elegant ways. But if nothing rings a bell, I guess I'll take the bullet and accept this as the best practical solution.